manim/mobject/probability.py

```diff
@@ -230,106 +213,218 @@ def construct(self):
-                height=(value / self.max_value) * self.total_bar_height,
-                width=buff,
+                height=bar_h,
+                width=bar_w,
                 stroke_width=self.bar_stroke_width,
                 fill_opacity=self.bar_fill_opacity,
             )
-            bar.move_to((2 * i + 1) * buff * RIGHT, DOWN + LEFT)
-            bars.add(bar)
-        bars.set_color_by_gradient(*self.bar_colors)
-
-        bar_labels = VGroup()
-        for bar, name in zip(bars, self.bar_names):
-            label = MathTex(str(name))
-            label.scale(self.bar_label_scale_val)
-            label.next_to(bar, DOWN, SMALL_BUFF)
-            bar_labels.add(label)
-
-        self.add(bars, bar_labels)
-        self.bars = bars
-        self.bar_labels = bar_labels
-
-    def change_bar_values(self, values):
-        for bar, value in zip(self.bars, values):
-            bar_bottom = bar.get_bottom()
-            bar.stretch_to_fit_height((value / self.max_value) * self.total_bar_height)
-            bar.move_to(bar_bottom, DOWN)
+            bar.next_to(self.c2p(i + 0.5, 0), UP, buff=0)
+            self.bars.add(bar)
+        if isinstance(self.bar_colors, str):
+            self.bars.set_color_by_gradient(self.bar_colors)
+        else:
+            self.bars.set_color_by_gradient(*self.bar_colors)
+
+        self.add_to_back(self.bars)
+
+    def add_x_labels(self):
+        if self.x_labels is not None or self.bar_names is None:
+            return
+        self.x_labels = VGroup()
+
+        max_lbl_width = 0
+        max_lbl_height = 0
+        for i, name in enumerate(self.bar_names):
+            if i == len(self.values):
+                self.bar_names = self.bar_names[:i]
+                break
+            label = self.x_label_constructor(name)
+            if max_lbl_width < label.width:
+                max_lbl_width = label.width
+            if max_lbl_height < label.height:
+                max_lbl_height = label.height
+            self.x_labels.add(label)
+
+        if self.x_label_scale_value is None:
+            unit_width = self.c2p(1, 0)[0] - self.c2p(0, 0)[0]
+            self.x_label_scale_value = min(unit_width * 0.75 / max_lbl_width, 0.75)
+
+        if self.x_label_buff is None:
+            self.x_label_buff = max_lbl_height * self.x_label_scale_value * 0.75
+
+        for i, label in enumerate(self.x_labels):
+            label.scale(self.x_label_scale_value)
+            label.move_to(
+                self.c2p(i + 0.5, 0)
+                + DOWN * max_lbl_height * self.x_label_scale_value / 2
+                + DOWN * self.x_label_buff
+            )
+
+        self.add(self.x_labels)
+
+    #
+    # def change_bar_values(self, values):
+    #     for bar, value in zip(self.bars, values):
+    #         bar_bottom = bar.get_bottom()
+    #         bar.stretch_to_fit_height((value / self.max_value) * self.total_bar_height)
+    #         bar.move_to(bar_bottom, DOWN)
```

Comment:
I forgot to delete this piece of code. I guess it's really useful to have a change_bar_value method so I'll be working on that